src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs

```diff
@@ -0,0 +1,103 @@
+    {
+        /// <summary>
+        ///     This is the list of test archives produced with OSPlatformConfig metadata set
+        /// </summary>
+        [Required]
+        public ITaskItem[] UnfilteredUploadItems { get; set; }
+
+        /// <summary>
+        ///     This is the TestTFM folder that we want to select test zips from. This needs to be on the path to the test zip.
+        /// </summary>
+        [Required]
+        public string FilterToTestTFM { get; set; }
+
+        /// <summary>
+        ///     This is the OSGroup folder that we want to select test zips from. This does not require an exact match as some
+        ///     tests
+        ///     may or may not have an exact OSGroup folder match.
+        /// </summary>
+        [Required]
+        public string FilterToOSGroup { get; set; }
+
+        /// <summary>
+        ///     This is the bin folder that should be removed from all paths.
+        /// </summary>
+        [Required]
+        public string BinDir { get; set; }
+
+        [Output]
+        public ITaskItem[] FilteredUploadItems { get; set; }
+
+        public override bool Execute()
+        {
+            List<ITaskItem> filteredBasedOnTestTfm =
+                UnfilteredUploadItems.Where(
+                    item => makeRelativePathTo(item.ItemSpec, BinDir).Contains(FilterToTestTFM)).ToList();
+
+            List<ITaskItem> localFilteredUploadItemsList = new List<ITaskItem>();
+
+            Dictionary<string, List<ITaskItem>> map = new Dictionary<string, List<ITaskItem>>();
+            foreach (var item in filteredBasedOnTestTfm)
+            {
+                string testName = Path.GetFileName(item.ItemSpec);
+                List<ITaskItem> values = null;
+                if (!map.TryGetValue(testName, out values))
+                {
+                    values = new List<ITaskItem>();
+                    map[testName] = values;
+                }
+                values.Add(item);
+            }
+
+            foreach (var item in map)
+            {
+                List<ITaskItem> values = item.Value;
+                ITaskItem selectedItem = null;
+                foreach (var value in values)
+                {
+                    if (value.GetMetadata("OSPlatformConfig").Contains(FilterToOSGroup))
+                    {
+                        selectedItem = value;
+                    }
+                }
+                if (selectedItem == null)
+                {
+                    selectedItem = values[0];
+                }
+                localFilteredUploadItemsList.Add(selectedItem);
+            }
+
+            FilteredUploadItems = localFilteredUploadItemsList.ToArray();
+            return true;
+        }
+
+
+        private string makeRelativePathTo(string path1, string path2)
+        {
+            path2 = Path.GetFullPath(path2);
+            string fullPath1 = Path.GetFullPath(path1);
+            if (fullPath1.StartsWith(path2))
```

Comment:
I agree that's probably ok, since we can't detect the case-sensitivity of the underlying OS.  In general, I think our build tries hard not to have thing meaningfully differ by case, so it's probably safe.